politeiad/backendv2/tstorebe/tstore/anchor.go

```diff
@@ -0,0 +1,569 @@
+// Copyright (c) 2020-2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package tstore
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"time"
+
+	dcrtime "github.com/decred/dcrtime/api/v2"
+	"github.com/decred/dcrtime/merkle"
+	"github.com/decred/politeia/politeiad/backendv2/tstorebe/store"
+	"github.com/google/trillian"
+	"github.com/google/trillian/types"
+	"google.golang.org/grpc/codes"
+)
+
+// TODO handle reorgs. An anchor record may become invalid in the case of a
+// reorg. We don't create the anchor record until the anchor tx has 6
+// confirmations so the probability of this occurring on mainnet is low, but it
+// still needs to be handled.
+
+const (
+	// anchorSchedule determines how often we anchor records. dcrtime
+	// currently drops an anchor on the hour mark so we submit new
+	// anchors a few minutes prior to that.
+	// Seconds Minutes Hours Days Months DayOfWeek
+	anchorSchedule = "0 56 * * * *" // At minute 56 of every hour
+
+	// anchorID is included in the timestamp and verify requests as a
+	// unique identifier.
+	anchorID = "tstorebe"
+)
+
+// anchor represents an anchor, i.e. timestamp, of a trillian tree at a
+// specific tree size. The LogRootV1.RootHash is the merkle root hash of a
+// trillian tree. This root hash is submitted to dcrtime to be anchored and is
+// the anchored digest in the VerifyDigest. Only the root hash is anchored, but
+// the full LogRootV1 struct is saved as part of an anchor record so that it
+// can be used to retrieve inclusion proofs for any leaves that were included
+// in the root hash.
+type anchor struct {
+	TreeID       int64                 `json:"treeid"`
+	LogRoot      *types.LogRootV1      `json:"logroot"`
+	VerifyDigest *dcrtime.VerifyDigest `json:"verifydigest"`
+}
+
+func (t *Tstore) droppingAnchorGet() bool {
+	t.Lock()
```

Comment:
This sounds very expensive for a mutex. Why is this ok to lock?

Reply:
This just returns the `droppingAnchor` bool that is used to verify that the anchor drop calls are not rentrant.  Its not expensive.